Use `ed25519::Signature` type; impl `Signer`/`Verifier` traits

tarcieri opened the pull request:
The `ed25519` crate provides a common `Signature` type which can be shared among libraries which implement Ed25519 signing/verification:

https://docs.rs/ed25519

Without common types/traits for Ed25519 signatures, downstream consumer crates which use Ed25519 as part of a high-level protocol implementation which allows consumers of their crates to be able to use their choice of Rust Ed25519 libraries wind up writing their own abstractions to do so on a crate-by-crate basis.

The `ed25519` crate, and the `signature` crate it's built on, provide a common abstraction layer:

- `ed25519` crate: https://github.com/RustCrypto/signatures/tree/master/ed25519
- `signature` crate: https://github.com/RustCrypto/traits/tree/master/signature

The `ed25519` crate has been natively adopted by `ed25519-dalek`:

- Issue: https://github.com/dalek-cryptography/ed25519-dalek/issues/80
- Usage example: https://docs.rs/ed25519/1.0.2/ed25519/#using-above-example-with-ed25519-dalek

If `sodiumoxide` were to adopt these types/traits as well, I'd be happy to add similar documentation to the `ed25519` crate for how to use them with `sodiumoxide`.

tarcieri commented on src/crypto/sign/ed25519.rs:
```diff
@@ -389,20 +402,21 @@ mod test {
-            round_trip(sig);
-        }
-    }
+    // TODO(tarcieri): enable `serde` feature
+    // #[cfg(feature = "serde")]
+    // #[test]
+    // fn test_serialisation() {
+    //     use randombytes::randombytes;
+    //     use test_utils::round_trip;
+    //     for i in 0..256usize {
+    //         let (pk, sk) = gen_keypair();
+    //         let m = randombytes(i);
+    //         let sig = sign_detached(&m, &sk);
+    //         round_trip(pk);
+    //         round_trip(sk);
+    //         round_trip(sig);
+    //     }
+    // }
```
Note: the `ed25519` crate has a `serde` feature which provides 64-byte serialized signatures in formats like bincode and CBOR.

I didn't wire it up yet in this PR since enabling a `serde` feature in a dependency when `serde` is already an optional crate dependency is a bit tricky. That said, we can probably use the same workaround I used in ed25519-dalek:

https://github.com/dalek-cryptography/ed25519-dalek/blob/952bdd0/Cargo.toml#L30

tarcieri commented:
FYI, I already maintain a "wrapper crate" which enables the usage of these Ed25519 signature types/traits:

https://docs.rs/signatory-sodiumoxide/0.21.0/signatory_sodiumoxide/

However, I think it'd be great if `sodiumoxide` added native support the same way `ed25519-dalek` did.

tarcieri commented:
Note: per #445 I'm definitely still interested in seeing this through. Just wanted to make sure there's actually interest in this PR before completing it.

kpp commented:
Yes I am interested in this PR.

tarcieri commented:
The tricky remaining question is what to do about `serde` support. If you'd like to use the built-in support in the `ed25519` crate, it'd require using a hack similar to the one in `ed25519-dalek`:

https://github.com/dalek-cryptography/ed25519-dalek/pull/124/files#r393865319

If that seems acceptable I can do the same thing, but I just wanted to make sure first.

Note that there's been quite a bit of discussion about `serde` support in `ed25519-dalek`, and it seems like the serializers the `ed25519` crate provides may differ from the ones you presently define. See this issue for some additional discussion:

https://github.com/dalek-cryptography/ed25519-dalek/pull/146

kpp commented:
TBO I personally can't find `serde` feature useful at all... However we can work on it.

tarcieri commented:
Okay, I can update the PR to do what I did in ed25519-dalek and we can go from there.

tarcieri commented:
I went ahead and rebased this PR, and for now just removed the `serde` test, as well as WIP.

I think it's pretty problematic to serialize Ed25519 signatures using `serde`. See the discussion on this issue:

https://github.com/dalek-cryptography/ed25519-dalek/pull/146

Without looping in an additional crate like `serde_bytes`, the serialization will be bad for at least one of: bincode, CBOR, JSON.

So I left it out for now, but if you'd like me to revisit that I can.

kpp commented:
@sodiumoxide/owners  How about breaking `serde` feature and revisiting in sometimes later when this feature is requested?

dnaq commented:
Agreed, as it stands now serde support is a misfeature and it will keep
being one until the derive-macros work on large arrays, but that will break
the serialization format anyway.
On Mon, 7 Dec 2020 at 18:06, Roman Proskuryakov <notifications@github.com>
wrote:

> @sodiumoxide/owners <https://github.com/orgs/sodiumoxide/teams/owners>
> How about breaking serde feature and revisiting in sometimes later when
> this feature is requested?
>
> —
> You are receiving this because you are on a team that was mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/sodiumoxide/sodiumoxide/pull/434#issuecomment-740051317>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJ5KBKMRDCWOTLNNQ2D6NLSTUDSRANCNFSM4RIRX2DQ>
> .
>

kpp commented:
bors: r+
